go/tools/gazelle/merger/merger.go

```diff
@@ -46,112 +46,300 @@ var (
 	if err != nil {
 		return nil, err
 	}
-	if err := shouldIgnore(f); err != nil {
+	if err := shouldIgnore(oldFile); err != nil {
 		return nil, err
 	}
 
+	oldStmt := oldFile.Stmt
 	var newStmt []bzl.Expr
-	for _, s := range newfile.Stmt {
-		c, ok := s.(*bzl.CallExpr)
+	for _, s := range genFile.Stmt {
+		genRule, ok := s.(*bzl.CallExpr)
 		if !ok {
-			return nil, fmt.Errorf("got %v expected only CallExpr in %q", s, newfile.Path)
+			return nil, fmt.Errorf("got %v expected only CallExpr in %q", s, genFile.Path)
 		}
-		other, err := match(f, c)
-		if err != nil {
-			return nil, err
-		}
-		if other == nil {
-			newStmt = append(newStmt, c)
+		i, oldRule := match(oldFile, genRule)
+		if oldRule == nil {
+			newStmt = append(newStmt, genRule)
 			continue
 		}
-		if name(c) == "load" {
-			mergeLoad(c, other, f)
+
+		var mergedRule bzl.Expr
+		if kind(oldRule) == "load" {
+			mergedRule = mergeLoad(genRule, oldRule, oldFile)
 		} else {
-			merge(c, other)
+			mergedRule = mergeRule(genRule, oldRule)
 		}
+		oldStmt[i] = mergedRule
 	}
-	f.Stmt = append(f.Stmt, newStmt...)
-	return f, nil
+
+	oldFile.Stmt = append(oldStmt, newStmt...)
+	return oldFile, nil
 }
 
-// shouldIgnore checks whether "gazelle:ignore" appears at the beginning of
-// a comment before or after any top-level statement in the file.
-func shouldIgnore(oldFile *bzl.File) error {
-	for _, s := range oldFile.Stmt {
-		for _, c := range s.Comment().After {
-			if strings.HasPrefix(c.Token, gazelleIgnore) {
-				return GazelleIgnoreError{c.Start}
+// merge combines information from gen and old and returns an updated rule.
+// Both rules must be non-nil and must have the same kind and same name.
+func mergeRule(gen, old *bzl.CallExpr) *bzl.CallExpr {
+	genRule := bzl.Rule{Call: gen}
+	oldRule := bzl.Rule{Call: old}
+	merged := *old
+	merged.List = nil
+	mergedRule := bzl.Rule{Call: &merged}
+
+	for _, k := range oldRule.AttrKeys() {
+		oldAttr := oldRule.Attr(k)
+		var mergedAttr bzl.Expr
+		if !mergeableFields[k] {
+			mergedAttr = oldAttr
+		} else {
+			genAttr := genRule.Attr(k)
+			var err error
+			mergedAttr, err = mergeAttr(genAttr, oldAttr)
+			if err != nil {
+				// TODO: add a verbose mode and log errors like this.
+				mergedAttr = genAttr
 			}
 		}
-		for _, c := range s.Comment().Before {
-			if strings.HasPrefix(c.Token, gazelleIgnore) {
-				return GazelleIgnoreError{c.Start}
+
+		if mergedAttr != nil {
+			mergedRule.SetAttr(k, mergedAttr)
+		}
+	}
+
+	for _, k := range genRule.AttrKeys() {
+		if mergedRule.Attr(k) == nil {
+			mergedRule.SetAttr(k, genRule.Attr(k))
+		}
+	}
+
+	return &merged
+}
+
+// mergeAttr combines information from gen and old and returns an updated attr.
+// The attrs may be strings, lists, a select call with a dict, or a list plus
+// a select call. An error is returned if the attrs can't be merged, for example
+// because they are in a format Gazelle doesn't understand.
+func mergeAttr(gen, old bzl.Expr) (bzl.Expr, error) {
```

Comment:
I think the fact that this is currently used to merge attributes is not a good reason to name it that.
In theory you could call this on anything that was an expression with a type that was understood (merging globals springs to mind)
I would just call it mergeExpression and doucment the expression types it works with.
The same is true of the other methods below.